driver/convert.c

```diff
@@ -1204,74 +1205,130 @@ static SQLRETURN double_to_binary(esodbc_rec_st *arec, esodbc_rec_st *irec,
-	size_t pos, octets;
-	/* buffer unit size */
-	size_t usize = wide ? sizeof(SQLWCHAR) : sizeof(SQLCHAR);
-	esodbc_state_et state = SQL_STATE_00000;
-	esodbc_stmt_st *stmt = arec->desc->hdr.stmt;
-	/* buffer is overprovisioned for !wide, but avoids double declaration */
-	SQLCHAR buff[(2 * ESODBC_PRECISION_INT64 + /*.*/1 + /*\0*/1)
-		* sizeof(SQLWCHAR)];
-	xstr_st xstr = (xstr_st) {
-		.wide = wide,
-		.c  = (cstr_st) {
-			/* same vals for both wide and C strings */
-			.str = buff,
-			.cnt = 0
-		}
-	};
+	static const char *flt_def = "%.*f";
+	static const wchar_t *flt_wdef = L"%.*f";
+	static const char *flt_exp = "%.*E";
+	static const wchar_t *flt_wexp = L"%.*E";
 
-	/*
-	 * split the whole and fractional parts
-	 */
-	assert(sizeof(dbl) == sizeof(whole)); /* [double]==[long long] */
-	whole = (long long)dbl;
-	rest = dbl - whole;
-
-	/* retain user defined or data source default number of fraction digits */
-	scale = 0 < arec->scale ? arec->scale : irec->es_type->maximum_scale;
-	rest *= pow10(scale);
-	rest = round(rest);
-	fraction = rest < 0 ? (unsigned long long) -rest
-		: (unsigned long long)rest;
-
-	/* copy integer part into work buffer */
-	pos = i64tot((int64_t)whole, buff, wide);
-	/* would writing just the whole part + \0 fit into the buffer? */
-	octets = buff_octet_size((pos + 1) * usize, usize, arec, irec, &state);
-	if (state) {
-		REJECT_AS_OOR(stmt, dbl, /*fixed?*/FALSE, "[STRING]<[floating.whole]");
+	esodbc_dbc_st *dbc = HDRH(stmt)->dbc;
+	double abs;
+
+	if (wide) {
+		switch (dbc->sci_floats) {
+			case ESODBC_FLTS_DEFAULT:
+				*rep = ESODBC_FLTS_DEFAULT;
+				return flt_wdef;
+			case ESODBC_FLTS_SCIENTIFIC:
+				*rep = ESODBC_FLTS_SCIENTIFIC;
+				return flt_wexp;
+			case ESODBC_FLTS_AUTO:
+				abs = (0 <= dbl) ? dbl : -dbl;
+				if (abs < 1E-3 || 1E7 <= abs) {
+					*rep = ESODBC_FLTS_SCIENTIFIC;
+					return flt_wexp;
+				} else {
+					*rep = ESODBC_FLTS_DEFAULT;
+					return flt_wdef;
+				}
+			default:
+				BUGH(stmt, "unexpected floats representation value: %d.",
+					dbc->sci_floats);
+				*rep = ESODBC_FLTS_DEFAULT;
+				return flt_wdef;
+		}
 	} else {
-		assert(octets == (pos + 1) * usize);
+		switch (dbc->sci_floats) {
+			case ESODBC_FLTS_DEFAULT:
+				*rep = ESODBC_FLTS_DEFAULT;
+				return flt_def;
+			case ESODBC_FLTS_SCIENTIFIC:
+				*rep = ESODBC_FLTS_SCIENTIFIC;
+				return flt_exp;
+			case ESODBC_FLTS_AUTO:
+				abs = (0 <= dbl) ? dbl : -dbl;
+				if (abs < 1E-3 || 1E7 <= abs) {
+					*rep = ESODBC_FLTS_SCIENTIFIC;
+					return flt_exp;
+				} else {
+					*rep = ESODBC_FLTS_DEFAULT;
+					return flt_def;
+				}
+			default:
+				BUGH(stmt, "unexpected floats representation value: %d.",
+					dbc->sci_floats);
+				*rep = ESODBC_FLTS_DEFAULT;
+				return flt_def;
+		}
 	}
+}
+
+static SQLRETURN double_to_str(esodbc_rec_st *arec, esodbc_rec_st *irec,
+	double dbl, void *data_ptr, SQLLEN *octet_len_ptr, BOOL wide)
+{
+#	define DBL_BASE10_MAX_LEN /*-0.*/3 + DBL_MANT_DIG - DBL_MIN_EXP
```

Comment:
I believe `DBL_MANT_DIG` and `DBL_MIN_EXP` refer to the base 2 representation, not base 10, so they're 53 and -1021 respectively.  The base 10 equivalents are 15 and -308.  In other words the number of characters to print the smallest `double` in fixed point notation would be 3 + 15 - -308, not 3 + 53 - -1021.  I think the correct constants in C99 are `DBL_DIG` and `DBL_MIN_10_EXP`.  But if Microsoft's headers don't define these or if you want to be allow a bit more in case there's some edge case where more digits are printed then maybe just define a constant to slightly higher like 336 (= 21 * 16) and use that for the buffer size including the `\0` terminator?

Reply:
Thanks for spotting that, fixed now.
The defs are available (https://docs.microsoft.com/en-us/cpp/cpp/floating-limits).
I've stuck with the "exact" calculation, since the buffer should anyways be way over-provisioned for the precision (=ODBC's `scale`) these numbers will be printed with. (Currently the limit is 19, but even that is probably too high -- pending further clarifications.)